### DISCUSSION TITLE
Join forces

### DISCUSSION THREAD

**Discussion by io-debug** [marq24/ha-senec-v3#6] [General]
Hi @marq24,

as you suggested, lets discuss how to join forces and how to merge our projects (https://github.com/io-debug/senecv3).
:-)

**Comment by marq24** [comment]
thanks for getting in contact with me - I really like your amount of documentation

**Reply by io-debug** [reply]
Thanks! When I have spare time, I'll have a closer look at your source code to understand the implementation.
Just as a simple suggestion: Let's use your your (this) project as basis for the future development.
What are your thoughts? (By the way, we also can change to German language, if you prefer)

**Reply by marq24** [reply]
thanks a lot for the PR! - I might not be that responsive in the next couple of days... The PR and any contribution is highly appreciate... 

And yes "internally" we can communicate in German (I am located in the Gütersloh Area) - but for the communication that is visible in public I prefer English (just to be as transparent as possible to everybody) - I hope that's fine with you.

**Comment by io-debug** [comment]
Is there anything you are currently working on? Just to align, that we dont work on the same topics/features in the next time. 

Topics/ideas I currently have in mind, but haven't analyzed yet:
- Config Flow / Option Flow: As a user I would like to select during setup which sensors should be activated and which should be deactivated.
- Spare capacity (Reservekapazität) of the Battery: As a user I would like to read and set the spare capacity to change this in automations e.g. based on the pv forecast or the weather.
- Peak Shaving: As a user I would like to read and set the peak shaving options.

**Comment by marq24** [comment]
- [x] Peak Shaving Options
- [x] Spare capacity handling (if possible?)

for the Activate/Deactivate Entities I am not so sure - I have to admit I have not seen this in any other integration I have personally used so far... In the current impl. I have set some Entities as "deactivated by default" - and if the user want it he can enable them via the regular GUI... [you can see details in in the const.py]... 

IMHO when you as user setting up the Integration you typically _do not have a single clue_ about the functionality and the general provided features - so from my point of view the user will not have enough information/knowledge what to choose and what not

**Comment by marq24** [comment]
Concerning the 'Spare capacity handling' and the current implemented version... I am no sure, if this is possible at all (to implement)...

In the current version we going to poll the current value each time (when we update also all the other sensor data)... Would it be possible to send these additional request only, a) if the number-sensor is active (enabled by the user)... and b) normally if you as user does not change this value, then it should not chnage that much during the day... so there might be a way to "lazy load" this number (e.g. only at startup of the integration - and then every 24h or so (if the user adjust it via the web)...

I would like to avoid unnecessary API requests to SENEC (they might going to dislike the fact, that we start a "denail of service attack" - obviously we don't do that - but it would be at least nice from our side, if we keep the traffic as low as possible.

What do you think? [I did not add this as comment to you PR by purpose - the current status is fine and I don't want to give you the impression, that I am asking for more and more and more... ]

**Reply by io-debug** [reply]
Great minds think alike.
I had a similar thought, but couldn't spend time on the analysis of the idea.
Here my first thoughts, how to implement:

-We should implement this for the Web Api as well as for the requests to the local Senec device to increase response time and reduce effort for the device/api
- We build buckets of sensors (e.g. one bucket with sensors that should be updated very often and one bucket with sensors that change very selden)
- We implement for each bucket an update-intervall
- Than the update coordinator has to "select" the sensors to update (not sure how to implement this without further analysis)

Regarding the current pull request:
-The changes are now commited to the development-branch of your git repository. Should we move the changes to the master-branch as well?

**Comment by marq24** [comment]
Bin heute Morgen etwas kurz angebunden... deswegen auch stumpf auf deutsch ;-)

Ich sehe zwei Themen [und zwar jetzt auch nur kurzfristig für das `update_spare_capacity`]

1) komplett Deaktivierung des pullens (des neuen URL) wenn Sensor nicht aktiviert - ich gehe davon aus, das man ein listener auf aktiviert/deaktiviert machen kann - oder aber noch einfacher es gibt eine Methode am sensor, die man überschreiben kann (super aufrufen) und dann ein zusätzlichen call zum MySenecWebPortal machen und dot ein boolean auf "true/false" setzen...

2) Polling nur alle 24h (oder auch 12h)... Das würde ich jetzt im ersten Schritt auch erstmal ganz stumpf lösen: 

- ein `last_polled_timestamp` - wenn die update methode aufgerufen wird, wird geschauft, ob "last_polled_timestamp + 24h" < now() - wenn true, dann  `last_polled_timestamp = now()` und den call zur neuen Funktion `update_spare_capacity` machen... Und dann natürlich noch - wenn man den Wert setzt, den `last_polled_timestamp` noch auf 0 setzen... damit beim nächsten update der neue Wert auch auf jeden Fall gelesen wird... 

....

Zu dem lokal polling (und da nur die Felder zu ziehen, die vom User auch aktiviert sind) - habe ich noch nicht wirklich eine Meinung - mir ist bislang nicht aufgefallen, dass bei (meinen) 5 Sek Interval sich der SENEC irgendwie quälen würde... [ich vermute mal aktuell frist es die meiste Zeit, dass wir felder anfragen, die es (aktuell) nicht git ?! (STATISTICS) und die "arme MCU" erst lange suchen muss, bis sie `VARIABLE_NOT_FOUND` auswirft..]

....
und noch zum Master Branch - da die HACS user auch "ohne" BETA Flag jederzeit den Master branch installieren können, würde ich den Merge in den Master erst dann machen, wenn 1 & 2 erfolgreich umgesetzt sind...

**Reply by io-debug** [reply]
Ich versuche mir die Punkte in den nächsten Tagen in Ruhe anzuschauen. Aktuell ist die Zeit bei mir leider auch sehr knapp.  :-)

**Comment by marq24** [comment]
I think I have a "working" solution for "active" or "disabled" (so issue 1) can be considered as done (if you agree) - pull development soon

**Comment by marq24** [comment]
also 2) have been implemented now... 

during testing I found something really strange - I will create an issue [and assign it to you](https://github.com/marq24/ha-senec-v3/issues/22) - please let me know, if you can confirm this behaviour - and if yes - WTF! - how to "fix" that?

**Comment by marq24** [comment]
and BTW - my max value of the spare capacity is 50%... [it's still funny that I am able to adjust this value]

**Comment by io-debug** [comment]
Thank you, for your time and effort.
I just got the first minutes in this week to look at your work and to test the changes.

1) Looking at the code, the changes look good.
2) The idea is great, but causes two side effects for me:

**Effect one:** 
- The integration updated the spare_capacity some seconds ago. I change the spare capacity. The dashboard will show the old spare capacity till the 24 hours are over. Here we have to forece a refresh and reset the timer.
_EDIT: I just see that you reset the timer in the set_spare_capacity. Strange. I'll look at this later as well._

**Effect two:**
- The integration updated the spare_capacity some seconds ago. Because of some reason (well yes this is more a theoretical case) I have to reload the integration. In this case the spare_capacity is set up with "0" as spare capacity.

I'll try to look at this today in the evening.

**Comment by marq24** [comment]
Re: Effect one:
Is that "not" working as expected for you? https://github.com/marq24/ha-senec-v3/blob/0eccc370ccf80fdf5f1f671619636b68699196e6/custom_components/senec/pysenec_ha/__init__.py#L1535 - we can even move this  into the `if res.status == 200:` block if you like...

my local testings this morning had been successful - BUT I do not have the slider in a dashboard - I just looked into the Integration istelf - where the "control" result is also displayed -> 

![image](https://github.com/marq24/ha-senec-v3/assets/8012359/f1b91d10-49c2-4207-b47a-4954dccc1d48)

when I change the slider - then after a short while the displayed % value (red area) will also get updated... and in my log I could see, that:
`[custom_components.senec.pysenec_ha] ***** update_spare_capacity(self) ********`

Re: Effect two:
IMHO when the Integration get's reloaded the __init__ method get's called and there (https://github.com/marq24/ha-senec-v3/blob/0eccc370ccf80fdf5f1f671619636b68699196e6/custom_components/senec/pysenec_ha/__init__.py#L1358) the timestamp get's reset too - but since I am a python noob I could be totally  wrong... 

> I'll try to look at this today in the evening.
I am not in a hurry either - take you time to review & test the changes I did last evening

**Reply by io-debug** [reply]
As promised, I now had the chance to spend more time to look at the code and test the functionallity. 

I don't know how to explain, but I was not able to reproduce the errors I've seen during my lunch break.
(I'm guess I was stupid and forgot to clear a cache during my quick test).

I implemented your idea to move ```self._QUERY_SPARE_CAPACITY_TS = 0```  after ```if res.status == 200:``` .
From my perspective this is more accurate, since the ```res.status == 200``` implies that the update of the spare capacity was successfull.
The Pull Request will follow in some minutes. 

So from my perspective the Update is ready to go.

Regarding #22 : I'll use the next hour to analyse the issue and share my results in the issue.

**Comment by marq24** [comment]
just last remaining question... can you set the spare capacity > 50% ?! -> I can't - if can't either, then we should set the max value also to 50...

**Reply by io-debug** [reply]
I'm fine to set the max value to 50%, since I cant set a higher value. Even not on mein-senec.de.
At the same time I suggest to review this when the "limited operation mode" (Konditionierungsbetrieb) is over.

Reson:
During my research I could not find any information in the Senec-Documents about the percentage that can be set as maximum.
Therefore I'm not sure if 50% is the max, or if this is a result of the "limited operation mode" (Konditionierungsbetrieb).

An update / Pull Request is on the way.

**Comment by io-debug** [comment]
Hey, 
after I had a lot to do at work, I'll try to be more active in the next days.

**_Updated Documentation**
In the first step I've created a new Pull Request for an updated README.md. (Some parts have to be reviewed and the Wallbox sensors added).

**_Peak Shaving**
In the next days I will start to think how to integrate the Peak Shaving Options.
Seems to be tricky at a first glance, but I have an Idea for a first iteration....

**Comment by marq24** [comment]
thanks for the update!

And totally fine, to spend time with other things - there is no pressure from any side - I do have also other 'hobbies: ;-)

**Comment by marq24** [comment]
... For the PeakSaving you have probably the need to set multiple fields/values in a single action - right? -> IMHO the best solution for this in HA is to provide a Service that can request show a form that allow the user to specify all the fields and then make a single call... 

I do this in the ha-waterkotte integration where you can specify some date fields - but the values in the backend are different attributes... Let me know if you want/need additional details

**Reply by io-debug** [reply]
Hey, 
seems as if you can read my mind. I' searching for an good way to set multiple values...
So thanks for the hint - I'll have a look at the Service. :-)

**Reply by io-debug** [reply]
Just a quick update:

I've  implemented new sensors for the Web API that use the same upate-logic as the spare capacity:
-Grid export limit (Einspeisebegrenzung)
-Peak shaving mode (off, manual, automatic)
-Peak shaving capacity limit (the current limit for the battery, if you use manual peak shaving mode)
-Peak shaving end time (end time, when the capacity limit ends in manual mode)

The new sensors are translated as well.
For this part of the implementation I just have to review the code.

I've also implemented the first version of a service to set the data.
![image](https://github.com/marq24/ha-senec-v3/assets/139538558/c562bdc7-6dae-44a3-adf0-fc0c9c2809d6)

My task for tomorrow: Write the code to set the data (provided by the service) via the webapi.

As soon as I have a first working version, I will commit the code for testing.

**Comment by marq24** [comment]
after I have merged the PR and adjusted minor things I have one remaining question (not sure if this is possible at all) when starting the service dialog - is it possible to use the "current" values of the sensors for each field?

**Comment by marq24** [comment]
Concerning the PeakShaving... would you mind trying to understand, how the Entity services are working?
https://developers.home-assistant.io/docs/dev_101_services/#entity-services
that might be even a better approach (this might also imply that we are able to show the current values of the sensor)... 

and then I have another remark/question - currently the new 4 sensors are in the Sensor Platform - should the two percentage values not should be read only "numbers"? and the mode can be a select? [this would list some of the fields in the control section]... 

both things are really just optional...

The ONLY reason why I did not published a release yet is, that I am STILL waiting that my PR at HACS gets processed and I have the feeling that with every new "master" push the PR get's also pushed back on the internal worklist... :-/

**Reply by io-debug** [reply]
Thank you for your Feedback Matthias!

**_Entity Services / Custom Services**
The Custom Services look like a good enhancement.
I'll definitely dig into this.

**_Current Peak Shaving Sensor implementation**
Your questions remind me to describe more detailed in a PR why I developed something the way I did or even better to share my thougths before implementing something, so that we can discuss, if there is a better solution.

I'm not sure if you have the Number Entity or the "device_class" / "native_unit_of_measurement" of the  SensorEntityDescription in your mind.
I've decided to use the Sensor Entity and not the Number Entity, since I dont wanted that a user can set a value in the first implementation. (And I understood that a number is an entity that allows the user to input an arbitrary value to an integration)
The resons: 

- When calling the Senec WebAPI the variables "Battery capacity limit" and "End time" will be ignored if the Peak Shaving mode is not "MANUAL". So I had two options in my mind. First Option: we have to make several calls to Senec (in the first place we send a request as if the user selected the "MANUAL" Peak Shaving mode to save the data and then set again the mode the user currently has) - This conflicts with our aim to reduce the calls to senec. Or the second option: we have to store the data in the background and send the variables when/if the user selects the "MANUAL" Peak Shaving mode later. When using the second option it could be irritating, because the user will not see updated data for the variable in the dashboard, since he selects the "MANUAL" Peak Shaving mode. Of cause we could "fake" this as well, but than we are showing wrong data in the case a user is not changing the Peak Shaving mode to "MANUAL". 

Talking about the device class I'm happy to use a different one, if there is one that is more suitable. 

Nevertheless I agree to your thougts.
I would love to make the peak shaving options better usable.
Here I was thinking about some ideas (not sure if possible):
-Implement a select sensor for the Peak Shaving mode and let the user select between "Deactivated" and "Automatic". Not "Manual", since you propably want or have to change the other values at the same time.
-Create one Sensor group that combines the functions to get/set data of the current sensors
-Connect the service (or entity service) with the sensors, so that you can set data from each of the sensors
-Provide a card for the frontend where you can see and set all Peak Shaving sensors at one place

What are your thoughts on this?

**_Release**
From my perspective there is no need to hurry. As you see I dont have as much time as I would love to, in order to participate...  :-)

HACS shows me the version 3.0.8-rc5 as the latest availabe version. Isn't this the latest version?

**Reply by marq24** [reply]
Good Morning David,

### Release
concerning HACS - of course when you have added this repo as custom, then all is fine - my goal is, to become part of the "build-in" HACS repositories - so I have created a PR @ HACS for this move... 

Right now the original source (of this fork) is a default repo - this means users will find a SENEC integration already in HACS - and when they try it the users will realize that it's not working... Which is obviously a bad user experience... So I want to add this repo also to the default HACS ones... 

### EntityServices
I have NO CLUE, if this can/will be successful - I just saw this in the docu when I searched for possible solutions to include the current entity values in the service ... I did NOT read anywhere, that a EntityService will solve this - I just found them and so that triggered my mind, if this can be a solution - so it could be a dead end! (In my internal tests I did not found a way to trigger an entity-service (they did not show up anywhere - but I just spend 1h with that...)

### Current Peak Shaving Sensor implementation
I am very happy with your PR - and actually (for me) there is no need to provide additional explanations in the PR (or here in a discussion in the first hand) - The NumberEntity just came to my mind - since it offers MIN and MAX values to be set... but of course (for the reasons you all mentioned) you don't want that the single entity is editable - so I was "hoping" for some sort of "read-only" flag within the entity configuration (somewhere) - but again - I did not searched yet myself...

So IMHO it's totally fine to start with an initial idea and implement that - if there are alternative thoughts/opinions/options we will share them with each other - and sometimes an suggestion is "valid" - but there are reasons not to follow it. My default approach is, (when I think something can be done different, that then I make the modification by myself and share it... But since in the moment I have three different integrations in active development it's just a question of time - and "here" we are already two :-D So I really appreciate everything you contribute!
